File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/Logger.java
package org.firstinspires.ftc.teamcode.util;

import org.firstinspires.ftc.robotcore.external.Telemetry;

public class Logger implements Log {
    private static final Logger instance = new Logger();
    private Telemetry telemetry = null;

    private Logger() {
    }

    public static Logger getInstance() {
        return instance;
    }

    public void setTelemetry(Telemetry telemetry) {
        this.telemetry = telemetry;
    }

    public void log(CAPTION caption, String format, Object... args) {
        if (null == telemetry) {
            System.out.println(String.format(format, args));
        } else {
            telemetry.addData(caption.name(), format, args);
            telemetry.update();
        }
    }

    @Override
    public void log(CAPTION caption, Object message) {
        if (null == telemetry) {
            System.out.println(message);
        } else {
            telemetry.addData(caption.name(), message);
            telemetry.update();
        }
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/baseClasses/Logger.kt
package org.firstinspires.ftc.teamcode.baseClasses

import com.acmerobotics.dashboard.FtcDashboard
import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import org.firstinspires.ftc.robotcore.external.Telemetry
import org.firstinspires.ftc.robotcore.external.matrices.OpenGLMatrix

class Logger {
    private var logs: MutableList<Pair<String, String>> = mutableListOf()
    private val telemetryPacket = TelemetryPacket()

    fun text(log: String) {
        val split = log.split("::")
        if (split.count() == 2) {
            logs.add(Pair(split[0].trim(), split[1].trim()))
            telemetryPacket.put(split[0].trim(), split[1].trim())
        } else {
            logs.add(Pair("", log))
            telemetryPacket.put("", log)
        }
    }

    fun text(title: String, value: String) {
        logs.add(Pair(title, value))
        telemetryPacket.put(title, value)
    }

    fun location(matrix: OpenGLMatrix) {
        val translation = matrix.translation
        val x = UnitDistance.mm(translation[0].toDouble()).inches
        val y = UnitDistance.mm(translation[1].toDouble()).inches

        val dotRadius = 10.0
        telemetryPacket.fieldOverlay()
                .setStroke("blue")
                .setFill("lightBlue")
                .setStrokeWidth(2)
                .fillCircle(x - dotRadius / 2, y - dotRadius / 2, dotRadius)
                .strokeCircle(x - dotRadius / 2, y - dotRadius / 2, dotRadius)
    }

    fun log(telemetry: Telemetry) {
        telemetry.clear()
        logs.forEach { telemetry.addData(it.first, it.second) }
        telemetry.update()
        FtcDashboard.getInstance().sendTelemetryPacket(telemetryPacket)
    }
}